tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,132 @@
+                               SystemMailboxesProvider systemMailboxesProvider, Clock clock) {
+        this.usersRepository = usersRepository;
+        this.jmapSettingsRepository = jmapSettingsRepository;
+        this.storeMailboxManager = storeMailboxManager;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash(RunningOptions runningOptions, CleanupContext context) {
+        return Flux.from(usersRepository.listReactive())
+            .transform(ReactorUtils.<Username, Task.Result>throttle()
+                .elements(runningOptions.getUsersPerSecond())
+                .per(Duration.ofSeconds(1))
+                .forOperation(username -> cleanupTrashForSingleUser(username, context)))
+            .reduce(Task.Result.COMPLETED, Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while accessing users from repository", e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username, CleanupContext context) {
+        return Mono.from(jmapSettingsRepository.get(username))
+            .flatMap(jmapSettings -> cleanupTrashForSingleUser(username, jmapSettings, context))
+            .defaultIfEmpty(Task.Result.COMPLETED)
+            .doOnNext(result -> {
+                LOGGER.info("Trash mailbox is cleaned for user {}", username);
+                context.incrementProcessed();
+            }).onErrorResume(e -> {
+                LOGGER.error("Error while cleaning trash for user {}", username, e);
+                context.addToFailedUsers(username.asString());
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username, JmapSettings jmapSettings, CleanupContext context) {
+        if (jmapSettings.trashCleanupEnabled()) {
+            return cleanupTrashForSingleUser(username,
+                toPeriod(jmapSettings.trashCleanupPeriod()),
+                context);
+        } else {
+            return Mono.just(Task.Result.COMPLETED);
+        }
+    }
+
+    private Period toPeriod(String trashCleanupPeriod) {
+        return MAP_PERIOD_SETTING_TO_DURATION.getOrDefault(trashCleanupPeriod, Period.ofMonths(1));
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username, Period period, CleanupContext context) {
+        return getMessageManagerForTrashMailbox(username)
+            .flatMap(messageManager ->
+                deleteExpiredMessages(getExpiredDate(period),
+                    context,
+                    messageManager,
+                    getMailboxSession(username)))
+            .then(Mono.just(Task.Result.COMPLETED));
+    }
+
+    private MailboxSession getMailboxSession(Username username) {
+        return storeMailboxManager.getSessionProvider().createSystemSession(username);
```

Comment:
Not fixed... As @chibenwa said, prefer `SessionProvider`

Meaning inject `SessionProvider` in the constructor instead of `StoreMailboxManager` (it's heavy to inject such a component just for getting a mailbox session)

Then just call `return sessionProvider.createSystemSession(username);`

Fix the test accordingly